proposed changes after first review (#33)
vcstuff/high-assurance-profile: draft-high-assurance-profile-oid4vc-sd-jwt-vc.md

.# Abstract

This document defines a profile of OpenID for Verifiable Credentials in combination with the credential format VC-SD-JWT. The aim is to select features and to define a set of requirements for the existing specifications to enable interoperability among Issuers, Wallets and Verifiers of Credentials where a high level of security and privacy is required. The profiled specifications include OpenID for Verifiable Credential Issuance [@!OIDF.OID4VCI], OpenID for Verifiable Presentations [@!OIDF.OID4VP], Self-Issued OpenID Provider v2 [@!OIDF.SIOPv2], and VC-SD-JWT [!I-D.ietf-terbu-sd-jwt-vc].

{mainmatter}

# Introduction

This document defines a set of requirements for the existing specifications to enable interoperability among Issuers, Wallets and Verifiers of Credentials where a high level of security and privacy is required. This document is an interoperability profile that can be used by implementations in various contexts, be it a certain industry or a certain regulatory environment.

> This document is not a specification, but a profile. It refers to specifications required for implementations to interoperate among each other and defines the set of features to be mandatory to implement, for the optionalities mentioned in the referenced specifications.

Reviewer: ```suggestion
This document is not a specification, but a profile. It refers to the specifications required for implementations to interoperate among each other and for the optionalities mentioned in the referenced specifications, defines the set of features to be mandatory to implement.
```

> The profile uses OpenID for Verifiable Credential Issuance [@!OIDF.OID4VCI] and OpenID for Verifiable Presentations [@!OIDF.OID4VP] as the base protocols for issuance and presentation of Credentials, respectively. The credential format used in the profile are W3C Verifiable Credentials secured using Selective Disclosure for JWTs (SD-JWT) according to the VC Data Model v2.0 [@VC-DATA], designated as VC-SD-JWTs. Additionally, considerations are given on how deployments can perform a combined issuance of credentials in both VC-SD-JWTs and ISO mdoc [@ISO.18013-5].

Reviewer: ```suggestion
The profile uses OpenID for Verifiable Credential Issuance [@!OIDF.OID4VCI] and OpenID for Verifiable Presentations [@!OIDF.OID4VP] as the base protocols for issuance and presentation of Credentials, respectively. The credential format used in the profile are W3C Verifiable Credentials secured using Selective Disclosure for JWTs (SD-JWT) according to the VC Data Model v2.0 [@VC-DATA], designated as VC-SD-JWTs. Additionally, considerations are given on how deployments can perform a combined issuance of credentials in both VC-SD-JWT and ISO mdoc [@ISO.18013-5] formats.
```

# Scope

The following aspects are in scope of this interoperability profile:

* Protocol for issuance of the Credentials online (can be both remote and in-person) (OID4VCI)
* Protocol for presentation of Credentials online (can be both remote and in-person) (OID4VP)

> * Protocol for User Authentication with the Wallet at a Verifier (SIOP v2)

Reviewer: ```suggestion
* Protocol for User Authentication by the Wallet at a Verifier (SIOP v2)
```

> * Wallet Attestation (during Credential issuance)

Author: because I think the aim is to attest the wallet, not a certain instance - An issuer needs to know it is talking to a trustworthy wallet (wallet provider/product), not a certain instance

> * Wallet Attestation (during Credential issuance)

Reviewer: ok..

|OID4VP| N/A |MUST|MUST|
|OID4VCI| MUST|MUST|N/A|
|SIOPv2|N/A|MUST|SHOULD|
|VC-SD-JWT|MUST|MUST|MUST|

# OpenID for Verifiable Credential Issuance

Implementations of this profile:

* MUST support both pre-auth code flow and authorization code flow.
* MUST support SD-JWT VC profile as defined in OID4VCI specification.

> * MUST support sender-constrained Tokens using a mechanism as defined in [@!I-D.ietf-oauth-dpop].

Reviewer: ```suggestion
* MUST support sender-constrained Tokens using a mechanism as defined in [@!I-D.ietf-oauth-dpop].

Both Wallet initiated and Issuer initiated issuance is supported.
```

## Credential Offer

* Both Grant Types `authorization_code` and `urn:ietf:params:oauth:grant-type:pre-authorized_code` MUST be supported as defined in Section 4.1.1 in [@!OIDF.OID4VCI]
* For Grant Type `authorization_code`, the Issuer MUST include a scope value in order to allow the Wallet to identify the desired credential type. The wallet MUST use that value in the `scope` Authorization parameter. For Grant Type `urn:ietf:params:oauth:grant-type:pre-authorized_code`, the pre-authorized code is used by the issuer to identify the credential type(s). (pending OID4VCI PR#519)
* As a way to invoke the Wallet, at least a custom URL scheme `haip://` MUST be supported. Implementations MAY support other ways to invoke the wallets as agreed by trust frameworks/ecosystems/jurisdictions, not limited to using other custom URL schemes.

> Note: The Authorization Code flow does not require a Credential Offer from the Issuer to the Wallet. However, it is included in the feature set of the Credential Offer because it might be easier to implement with existing libs and on top of existing implementations than the pre-authorized code Grant Tye.

Reviewer: ```suggestion
Note: The Authorization Code flow does not require a Credential Offer from the Issuer to the Wallet. However, it is included in the feature set of the Credential Offer because it might be easier to implement with existing libraries and on top of existing implementations than the pre-authorized code Grant Type.
```

   * Refresh tokens MUST be supported for credential refresh 
   * Wallets MUST be able to process deferred authorization, i.e. processing of the Token error response parameters `authorization_pending` and `slow_down`, and the credential offer parameter `interval` MUST be supported.
   * The wallet attestation JWT scheme is defined in (#wallet-attestation-schema)

Note: It is RECOMMENDED to use ephemeral client attestation JWTs for client authentication in order to prevent linkability across Credential Issuers.

Note: Issuers should be mindful of how long the usage of the refresh token is allowed to refresh a credential, as opposed to starting the issuance flow from the beginning. For example, if the User is trying to refresh a credential more than a year after its original issuance, the usage of the refresh tokens is NOT RECOMMENDED.

### Wallet Attestation Schema {#wallet-attestation-schema}

[Section 3.1 of wallet attestation draft would define the basics, and this profile will define the details.]

## Credential Endpoint

   * The `JWT` proof type MUST be supported.

## Server Metadata

> The Credential Issuer MUST publish a mapping of every Credential Type it supports to a scope value. 

Reviewer: ```suggestion
* The Credential Issuer MUST publish a mapping of every Credential Type it supports to a scope value.
```

   * Presentation Definition JSON object MUST be sent using a `presentation_definition` parameter.
   * The following features from the DIF Presentation Exchange v2.0.0 MUST be supported. A JSON schema for the supported features is in (#presentation-definition-schema):

    * In the `presentation_definition` object, `id`, `input_descriptors` and `submission_requirements` properties MUST be supported.
    * In the `input-descriptors` object, `id`, `name`, `purpose`, `group`, `format`, and `constraints` properties MUST be supported. In the `constraints` object, `limit_disclosure`, and `fields` properties MUST be supported. In the `fields` object, `path` and `filter` properties MUST be supported. A `path` MUST contain exactly one entry with a static path to a certain claim. A `filter` MUST only contain `type` elements of value `string` and `const` elements.
    * In the `submission_requirements` object, `name`, `rule (`pick` only)`, `count`, `from` properties MUST be supported.

# Self-Issued OP v2

To authenticate the user, subject identifier in a Self-Issued ID Token MUST be used as defined in [@!OIDF.SIOPv2].

>    * As a way to invoke the Wallet, at least a custom URL scheme `haip:// MUST` be supported. Implementations MAY support other ways to invoke the wallets as agreed by trust frameworks/ecosystems/jurisdictions, not limited to using other custom URL schemes.

Reviewer: ```suggestion
   * As a way to invoke the Wallet, at least a custom URL scheme `haip://` MUST be supported. Implementations MAY support other ways to invoke the wallets as agreed by trust frameworks/ecosystems/jurisdictions, not limited to using other custom URL schemes.
```

   * `subject_syntax_types_supported` value MUST be `urn:ietf:params:oauth:jwk-thumbprint`

# SD-JWT VCs {#sd-jwt-vc}

> As credential format, VC-SD-JWT VC as defined in [!I-D.ietf-terbu-sd-jwt-vc] MUST be used.

Reviewer: ```suggestion
As credential format, VC-SD-JWT as defined in [!I-D.ietf-terbu-sd-jwt-vc] MUST be used.
```

* It is at the discretion of the Issuer whether to use `exp` claim or a `status` claim to express the validity period of an SD-JWT-VC. The wallet and the verifier  MUST support both mechanisms.
* `iss` claim MUST be used to obtain Issuer’s signing key as defined in (#issuer-key-resolution). `iss` claim MUST be either an HTTPS URL or a Subject Alternative Name (SAN) from an X.509 certificate. Wallets and the Verifiers MUST support both options.
* `type` JWT claim as defined in draft-terbu-sd-jwt-vc. For details on the `cnf` claim, see (#holder-key-resolution).
* `status` claim MUST be present as defined in draft-terbu-sd-jwt-vc/draft-looker-jwt-cwt-status-list
* For Cryptographic Holder Binding, HB-JWT MUST be supported as defined in [@!I-D.ietf-oauth-selective-disclosure-jwt].
* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT-VC. The wallet and the verifier  MUST support both mechanisms.
* The `iss` claim MUST be an HTTPS URL. The `iss` value is used to obtain Issuer’s signing key as defined in (#issuer-key-resolution).
* The `type` JWT claim as defined in [!I-D.ietf-terbu-sd-jwt-vc].

> Further claims, either registered or private JWT claims, can be added to the credential as required by the respective credential type, or determined by the Issuer or the Holder. All claims that are not understood by implementations MUST be ignored, as defined in section 4, [@!RFC7519].

Reviewer: I think this is an important note?

Reviewer: ok, opened an issue/PR in sd-jwt-vc and an issue #37 in this repo

> Note: this requirement might change when claim-based and biometrics-based holder binding become available.

Reviewer: this seems duplicative with line 192. how about the order of 194 -> 192&196 consolidated?

Author: 192 is "It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT-VC. The wallet and the verifier  MUST support both mechanisms". I don't understand.

Reviewer: the line numbers moved because of the changes accepted. with the new lines, it will be 198 -> 196&200 consoludated.

Author: reworked cnf/holder binding text

Note: Re-using the same Credential across Verifiers, or re-using the same JWK value across multiple Credentials gives colluding Verifiers a mechanism to correlate the User. There are currently two known ways to address this with SD-JWT VCs. First is to issue multiple instances of the same credentials with different JWK values, so that if each instance of the credential is used at only one Verifier, it can be reused multiple times. Another is to use each credential only once (ephemeral credentials). It is RECOMMENDED to adopt one of these mechanisms.

Note: If there is a requirement to communicate information about the verification status and identity assurance data of the claims about the subject, the syntax defined by @!OIDF.ekyc-ida SHOULD be used. It is up to each jurisdiction and ecosystem, whether to require it to the implementers of this profile.

> Note: If there is a requirement to provide the Subject’s identifier assigned and maintained by the Issuer, `sub` claim MAY be used. There is no requirement for a binding to exist between `sub` and `cnf` claims. See section X in @!I-D.draft-terbu-sd-jwt-vc for implementation considerations.

Reviewer: ```suggestion
Note: If there is a requirement to provide the Subject’s identifier assigned and maintained by the Issuer, `sub` claim MAY be used. There is no requirement for a binding to exist between `sub` and `cnf` claims. See section X in [!I-D.draft-terbu-sd-jwt-vc] for implementation considerations.
```

Note: In some credential types, it is not desirable to include an expiration date (eg: diploma attestation). Therefore, this profile leaves its inclusion to the Issuer, or the body defining the respective credential type.

## Issuer identification and key resolution to validate an issued Credential {#issuer-key-resolution}

> This profile supports two ways to represent and resolves the key required to validate the issuer signature of a SD-JWT-VC, web-based key resolution and x.509 certificates.

Reviewer: ```suggestion
This profile supports two ways to represent and resolves the key required to validate the issuer signature of a VC-SD-JWT, web PKI-based key resolution and x.509 certificates.
```

Author: Why would you call it web PKI? I think web based key resolution is sufficient.

Reviewer: it's trust in the DNS sytem, that allows to trust the .well-known path. i thought it should be clearer

* Web-based key resolution: The key used to validate the Issuer’s signature on the VC-SD-JWT MUST be obtained from the VC-SD-JWT issuer's metadata as defined in Section 5 of [!I-D.ietf-terbu-sd-jwt-vc]. The JOSE header `kid` MUST be used to identify the respective key.
* x.509 certificates: the VC-SD-JWT contains the issuer's certificate along with a trust chain in the `x5c` JOSE header. In this case, the `iss` value MUST be an URL with a FQDN matching a `dNSName` Subject Alternative Name (SAN) [@!RFC5280] entry in the leaf certificate.

Note: The issuer MAY decide to support both options. In which case, it is at the discretion of the Wallet and the Verifier which key to use for the issuer signature validation.

### Cryptographic Holder Binding between VC and VP

> * For Cryptographic Holder Binding, an HB-JWT as defined in [!I-D.ietf-terbu-sd-jwt-vc] MUST always be present when presenting a VC-SD-JWT.

Reviewer: I think we can remove this section (lines 215-217) since now HB JWT is defined in SD-JWT spec itself and oliver's draft mandates HB jwt?

Reviewer: oh wow, I thought we have agreed to make HB-JWT mandatory in sd-jwt vc. opened an [issue #110](https://github.com/vcstuff/draft-terbu-sd-jwt-vc/issues/110) there and here - issue #38
